### ISSUE TITLE
Wrong handling of term ordering leads to incorrect differentiation

### ISSUE THREAD

**Issue by e-dorigatti** [matthewwardrop/formulaic#235]
Hello, differentiation in the latest version sometimes gives incorrect results compared to the earlier version:

```
import formulaic
print(formulaic.__version__)

f = formulaic.Formula("x1 + x2 + x3 + x1:x2 + {x3**2}")
print(f)

for var in ['x1', 'x2', 'x3']:
    print(var, f.differentiate(var, use_sympy=True))
```

Here are the outputs:

```
1.1.0
1 + x1 + x2 + x3 + x3 ** 2 + x1:x2
x1 0 + 1 + 0 + 0 + 0 + x2
x2 0 + 0 + 1 + 0 + 0 + x1
x3 0 + 0 + 0 + 1 + 0 + (2*x3)
```

and

```
1.0.2
1 + x1 + x2 + x3 + x3 ** 2 + x1:x2
x1 root:
    [0, 1, 0, 0, 0, x2]
x2 root:
    [0, 0, 1, 0, 0, x1]
x3 root:
    [0, 0, 0, 1, (2*x3), 0]
```

Seems like the issue is related to the default ordering of the terms, as the correct results can be obtained by explicitly disabling it:

```
f = formulaic.Formula("x1 + x2 + x3 + x1:x2 + {x3**2}")
f.ordering = formulaic.formula.OrderingMethod.NONE
```

**Comment by matthewwardrop** [comment]
Hi @e-dorigatti ! Thanks for surfacing this! I'll make sure it gets fixed soon.

I didn't realise anyone other than myself had ever shown an interest in the differentiation features. How do you use them in practice?

**Comment by e-dorigatti** [comment]
Hi @matthewwardrop thank you for your prompt answer! We are using formulas in our [bofire](https://github.com/experimental-design/bofire) package to express the model for an experimental design, and use the differentiation features to optimize the design with respect to some optimality criteria. In practice we use this to plan experiments :)

**Comment by matthewwardrop** [comment]
Nice! Maybe I should upgrade them from experimental status then :).

**Comment by matthewwardrop** [comment]
This has been fixed and released as 1.1.1.

**Comment by jiboncom** [comment]
Just wanted to mention that the differentiation methods are great!
